prevent-broker-scale-down-if-contains-partition-replicas.md

```diff
@@ -0,0 +1,82 @@
+  Changing the Kafka replica count directly in the Kafka Cluster would help us ensure that we keep same replicas everywhere like while generating certificates, services, ingresses, routes etc.
+- The broker certificates, services, ingresses, routes etc. will be treated with the original number of nodes and the rest of the reconciliation will be done normally.
+- We also generate a new condition which will be added to Kafka resource status depicting that the scale down is not done. It will also contain the `spec.replicas` count(which is being currently being used) in the condition message.
+  ```yaml
+  status:
+    clusterId: DoRj5f84Sruq_7TJ31y7Zw
+    conditions:
+      - lastTransitionTime: "2023-02-22T10:18:56.578009768Z"
+        message: 'Cannot Scale down since broker contains partition replicas. The `spec.kafka.replicas` should be reverted back to 4 directly in the Kafka resource'.
+        reason: ScaleDownException
+        status: "True"
+        type: Warning
+      - lastTransitionTime: "2023-02-22T10:18:57.664668863Z"
+        status: "True"
+        type: Ready
+  ```
+  Note: By the time the replicas are reverted back, the storage validation will be already complete based on the replica count present in Kafka custom resource.
+  This can can cause some issues if someone tries to make some forbidden changes (changes that might not be supported) to the storage during this time frame.
+  This is hard to prevent in the current code.
+  But the likelihood of this happening at the same time should be relatively small.
+  So this proposal suggests to ignore this risk.
+
+### How to bypass the broker scale down mechanism
+
+- To bypass the broker scale down mechanism you can use the annotation `strimzi.io/bypass-broker-scaledown-check: "true"` on the Kafka custom resource. For e.g.
+  ```sh
+  kubectl annotate Kafka my-cluster strimzi.io/bypass-broker-scaledown-check: "true"
+  ```
+
+### Other Scenarios
+
+- During the check, if the admin client is not able to connect to the cluster (not able to get the topic details), we will update the status of the Kafka CR with the respective warning and revert back the replica count in the `KafkaCluster` class.
+- If the Kafka cluster is just initialized and the pods are not ready, the `canScaleDownBrokers()` utility method will not work because the current Kafka replicas (replicas before the Kafka custom resource is modified) count will be equal to 0 and the Kafka replicas present in the Kafka custom resource will also be 0. Hence, the mechanism will not run since the condition requires the current replica count to be greater than the Kafka replicas present in the Kafka custom resource
+- If the current Kafka replicas/pods are 0 the mechanism will not work since if there are zero brokers it will not be considered a scaledown.
```

Comment:
```suggestion
- If the current Kafka replicas/pods are 0, the mechanism will not work since if there are zero brokers it will not be considered a scale-down.
```